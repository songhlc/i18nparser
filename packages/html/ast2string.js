import * as html from 'html5parser'
function ast2String (ast, text) {
  if (!text) {
    text = ''
  }
  ast.forEach(item => {
    if (item.type === html.SyntaxKind.Text) {
      text += item.value
    } else if (item.type === html.SyntaxKind.Tag) {
      text += '<' + item.name
      item.attributes && item.attributes.forEach(attr => {
        if (attr.value)
          text += ' ' + attr.name.value + '=' + attr.value.quote + attr.value.value + attr.value.quote
      })
      if (item.name.toLowerCase() == '!doctype') {
        text += ' html'
      }
      text += '>'
      if (item.body) {
        text = ast2String(item.body, text)
      }
      if (item.close) {
        text += item.close.value
      }
    }
  })
  return text
}
export default ast2String
