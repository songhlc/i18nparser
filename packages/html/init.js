import ast2string from './ast2string'
import * as html from 'html5parser'
import * as recast from "recast"
import { needtraslate } from '../utils'
var isInScript = false
/**
 * 
 * @param {*} input 
 * @param {
 *  tagEnter: Function, 
 *  scriptEnter: Function,
 *  textEnter: Function,
 *  leave: Function
 * } option 
 */
function init (input, options) {
  const ast = html.parse(input);
  const { tagRule, scriptRule, textRule, leave } = options
  html.walk(ast, {
    enter: (node) => {
      if (node.type === html.SyntaxKind.Tag) {
        if (node.name == "script") {
          isInScript = true
        }
        tagRule && tagRule.forEach(rule => {
          rule(node)
        })
      } else {
        // html中解析内部js
        if (isInScript) {
          const jsast = recast.parse(node.value);
          scriptRule && scriptRule.forEach(rule => {
            rule(jsast)
          })
          // js转换后重新解析
          const output = recast.print(jsast).code;
          node.value = output
        }
        if (needtraslate(node?.value)) {
          textRule && textRule.forEach(rule => {
            rule(node)
          })
        }
      }
    },
    leave: (node) => {
      if (isInScript) {
        isInScript = false
      }
    }
  });
  return ast
}
export default init