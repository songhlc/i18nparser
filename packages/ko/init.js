import { init, ast2string } from '../html'
import rules from '../rules/html'
export default (input) => {
    const { koTagRule, koTextRule } = rules
    var ast = init(input, {
        tagRule: koTagRule,
        textRule: koTextRule
      })
    return ast2string(ast)
}


