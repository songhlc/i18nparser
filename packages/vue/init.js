// 把.vue文件拆分成template、script和style
// tempate按照html来处理，script按照js来处理，style不处理
import { init as htmlinit, ast2string } from '../html'
import htmlrules from '../rules/html'
import scriptinit from '../rules/script'
import * as html from 'html5parser'
const { vueTagRule, vueTextRule } = htmlrules
var init = function (input) {
  const ast = html.parse(input);
  var isInScript = false
  var templateNode = null
  var scriptNode = null
  var styleNode = null
  templateNode = ast.filter(node => {
    return node.name == "template"
  })
  scriptNode = ast.filter(node => {
    return node.name == "script"
  })
  styleNode = ast.filter(node => {
    return node.name == "style"
  })
  // var htmlAst = htmlinit(ast2string(templateNode), { tagRule: vueTagRule, textRule: vueTextRule })
  var scriptAst = scriptinit(scriptNode[0].body[0].value)
  // console.log(ast2string(htmlAst))
  console.log(scriptAst)
  // console.log(ast2string(styleNode))
}
export default init