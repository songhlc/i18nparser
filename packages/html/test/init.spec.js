import init from '../init'
import * as html from 'html5parser'
import ast2string from '../ast2string'
const input = ` <!doctype html>
    <html>
      <body>
        <h1 id="hello">Hello world</h1>
        <h2 title="有没有翻译">中文翻译</h2>
        <h3 data-bind="text: '你好'">世界</h3>
        <script>
          window.alert('error')
          function aaa (p) {
            var b = p + '1'
            console.log(b);
          } 
        </script>
      </body>
    </html>
    `;
describe('测试init', () => {
  // 这是mocha的玩法，jest可以直接兼容
  it('测试空tagRule，空scriptRule，空textRule', () => {
    var ast = init(input, {
      tagRule: [
        node => {
        }
      ],
      scriptRule: [
        jsast => {
        }
      ],
      textRule: [
        node => {
        }
      ]
    })
    var result = ast2string(ast)
    expect(result == input).toBe(true);
    expect(result).toMatchSnapshot()
  })
  it('测试textRule', () => {
    var ast = init(input, {
      textRule: [
        node => {
          node.value = 'textrule规则'
        }
      ]
    })
    var result = ast2string(ast)
    expect(result == input).toBe(false);
    expect(result).toMatchSnapshot()
  })
  it('测试tagRule 找到data-bind 添加一个自定义属性data-i18nparser', () => {
    var ast = init(input, {
      tagRule: [
        node => {
          if (node.attributes.filter(attr => {
            return attr?.name?.value == 'data-bind'
          }).length > 0) {
            node.attributes.push({
              name: {
                value: "data-i18nparser",
                type: "Text"
              },
              value: {
                "quote": '"',
                "value": "i18nparser"
              }
            })
          }

        }
      ]
    })
    var result = ast2string(ast)
    expect(result == input).toBe(false);
    expect(result).toMatchSnapshot()
  })
})
