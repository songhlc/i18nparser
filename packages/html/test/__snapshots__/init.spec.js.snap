// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试init 测试tagRule 1`] = `
" <!doctype html>
    <html>
      <body>
        <h1 id=\\"hello\\">Hello world</h1>
        <h2 title=\\"有没有翻译\\">中文翻译</h2>
        <h3 data-bind=\\"text: '你好'\\" data-i18nparser=\\"i18nparser\\">世界</h3>
        <script>
          window.alert('error')
          function aaa (p) {
            var b = p + '1'
            console.log(b);
          } 
        </script>
      </body>
    </html>
    "
`;

exports[`测试init 测试textRule 1`] = `
" <!doctype html>
    <html>
      <body>
        <h1 id=\\"hello\\">Hello world</h1>
        <h2 title=\\"有没有翻译\\">textrule规则</h2>
        <h3 data-bind=\\"text: '你好'\\">textrule规则</h3>
        <script>
          window.alert('error')
          function aaa (p) {
            var b = p + '1'
            console.log(b);
          } 
        </script>
      </body>
    </html>
    "
`;

exports[`测试init 测试空tagRule，空scriptRule，空textRule 1`] = `
" <!doctype html>
    <html>
      <body>
        <h1 id=\\"hello\\">Hello world</h1>
        <h2 title=\\"有没有翻译\\">中文翻译</h2>
        <h3 data-bind=\\"text: '你好'\\">世界</h3>
        <script>
          window.alert('error')
          function aaa (p) {
            var b = p + '1'
            console.log(b);
          } 
        </script>
      </body>
    </html>
    "
`;
